graphistry/embed_utils.py

```diff
@@ -0,0 +1,593 @@
+        pbar = trange(epochs, desc=None)
+        model.to(device)
+
+        score = 0
+        for epoch in pbar:
+            model.train()
+            for data in g_dataloader:
+                g, edges, labels = data
+
+                g = g.to(device)
+                edges = edges.to(device)
+                labels = labels.to(device)
+
+                emb = model(g)
+                loss = model.loss(emb, edges, labels)
+                optimizer.zero_grad()
+                loss.backward()
+                nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
+                optimizer.step()
+                pbar.set_description(f"epoch: {epoch+1}, loss: {loss.item():.4f}, score: {score:.4f}%")
+
+            model.eval()
+            self._embeddings = model(self.g_dgl.to(device)).detach()
+            self._embed_model = model
+            if self._eval_flag:
+                score = self._eval(threshold=0.5)
+                pbar.set_description(f"epoch: {epoch+1}, loss: {loss.item():.4f}, score: {score:.2f}%")
+
+        return self
+
+    def embed(
+        self,
+        relation,
+        proto='DistMult',
+        embedding_dim=32,
+        use_feat=False,
+        X=None,
+        epochs=2,
+        batch_size=32,
+        train_split=0.8,
+        lr=1e-2,
+        inplace=False,
+        device="cpu",
+        evaluate=True,
+        *args,
+        **kwargs
+    ):
+
+        """Embed a graph using a relational graph convolutional network (RGCN),
+            and return a new graphistry graph with the embeddings as node
+            attributes.
+
+        Args:
+            relation pd.column: column to use as relation between nodes
+            proto (str, optional): metric to use, ['TransE', 'RotateE',
+                'DistMult'] or provide your own. Defaults to 'DistMult'.
+            d (int, optional): relation embedding dimension. Defaults to 32.
+            use_feat (bool, optional): whether to featurize nodes, if False
+                will produce random embeddings and shape them during training.
+                Defaults to True.
+            X (List or pd.DataFrame, optional): Which columns in the nodes
+                dataframe to featurize. Inherets args from
+                graphistry.featurize()
+                Defaults to None.
+            lr (float, optional): learning rate. Defaults to 0.003.
+            evaluate (bool, optional): whether to evaluate the model.
+                Defaults to False.
+            epochs (int, optional): traing epoch. Defaults to 2.
+            batch_size (int, optional): batch size. Defaults to 32.
+            train_split (float, optional): train percentage, between 0, 1.
+                Defaults to 0.8.
+
+        Returns:
+            self: graphistry instance
+        """
+        if inplace:
+            res = self
+        else:
+            res = self.bind()
+
+        res._relation = relation
```

Comment:
should all be top-level (reused in non-embed code), or maybe some can be put into an `._embed_opts` ?